### DISCUSSION TITLE
Hotwire support

### DISCUSSION THREAD

**Discussion by amo13** [livebud/bud#113] [Ideas]
Thank you a lot for this amazing project! Finally something actually viable for go.

I'd like to suggest considering to add something like [hotwire](https://hotwired.dev/). It is the new default in Rails and its integration seems to make a lot of sense.

**Comment by zeventhal** [comment]
I was thinking that default gRPC support would work better, gRPC manages sockets, and allows for type safe streams, if there is a go library that can compile go structs into .proto code then it would be a big addition to the framework and would greatly improve the frontend to backend communication.

**Reply by matthewmueller** [reply]
AFAIK gRPC is still not really supported in the browser.

> greatly improve the frontend to backend communication.

I definitely want to address this. The current plan is to code generate Typescript definitions based on the controller API that the views reference to ensure type-safety across the network boundary.

**Comment by matthewmueller** [comment]
Hey @amo13, thanks for your interest in Bud! I wrote something in [a HN comment](https://news.ycombinator.com/item?id=31372024), so I'll pull that in here since the current plan hasn't changed:

> Any plans to implement something like Phonenix LiveView? I've never used Go, but from what I've read I suppose it should work really well given Go's concurrency model. (The name "Livebud" gave some hope as well.)

I haven't looked too closely at LiveView yet. I did flip back and forth between going with the Hotwire approach (I guess inspired by LiveView?). It was a tough call, but I ultimately decided to server-render JS. This thought-experiment helped:

- If Ruby could natively render Javascript, would they still choose the Hotwire approach? I concluded no. I don't see the reason to have multiple approaches (HTML over wire, Stimulus), when one solution (SSR + hydration) would suffice. The one solution also happens to be way more familiar to today's frontend developers. I definitely didn't want to make another templating language.

Bud will eventually ship with it's own JS evaluator. Either through https://github.com/dop251/goja or through porting https://github.com/bellard/quickjs to Go (~25kloc of C)
